pkg/api/common/helper.go

```diff
@@ -337,3 +340,268 @@ func WrapAsParameter(s string) string {
 func WrapAsVerbatim(s string) string {
 	return fmt.Sprintf("',%s,'", s)
 }
+
+// GetDockerConfig transforms the default docker config with overrides. Overrides may be nil.
+func GetDockerConfig(opts map[string]string, overrides []func(*DockerConfig) error) (string, error) {
+	config := GetDefaultDockerConfig()
+
+	for i := range overrides {
+		if err := overrides[i](&config); err != nil {
+			return "", err
+		}
+	}
+
+	dataDir, ok := opts[ContainerDataDirKey]
+	if ok {
+		config.DataRoot = dataDir
+	}
+
+	b, err := json.MarshalIndent(config, "", "    ")
+	return string(b), err
+}
+
+// GetContainerdConfig transforms the default containerd config with overrides. Overrides may be nil.
+func GetContainerdConfig(opts map[string]string, overrides []func(*ContainerdConfig) error) (string, error) {
+	config := GetDefaultContainerdConfig()
+
+	for i := range overrides {
+		if err := overrides[i](&config); err != nil {
+			return "", err
+		}
+	}
+
+	dataDir, ok := opts[ContainerDataDirKey]
+	if ok {
+		config.Root = dataDir
+	}
+
+	buf := new(bytes.Buffer)
+	err := toml.NewEncoder(buf).Encode(config)
+	return buf.String(), err
+}
+
+// ContainerdKubenetOverride transforms a containerd config to set details required when using kubenet.
+func ContainerdKubenetOverride(config *ContainerdConfig) error {
+	config.Plugins.IoContainerdGrpcV1Cri.CNI.ConfTemplate = "/etc/containerd/kubenet_template.conf"
+	return nil
+}
+
+// ContainerdSandboxImageOverrider produces a function to transform containerd config by setting the SandboxImage.
+func ContainerdSandboxImageOverrider(image string) func(*ContainerdConfig) error {
+	return func(config *ContainerdConfig) error {
+		config.Plugins.IoContainerdGrpcV1Cri.SandboxImage = image
+		return nil
+	}
+}
+
+// DockerNvidiaOverride transforms a docker config to supply nvidia runtime configuration.
+func DockerNvidiaOverride(config *DockerConfig) error {
```

Comment:
this can also be a template snippet?

Reply:
This applies transformations to the struct, which are serialized all at once. I tried doing everything in the template in my original PR, but it was very difficult to guarantee correct serialization. 

So I converted everything to structs and serialize it once, and unit test against the serialization for both cases.

Reply:
not sure I understand the issue, if it's just template then the unit test just compare the templates. why do we need serialization?

Reply:
Why allow for humans to make errors in whitespace? I made several typos in the original PR and aks-engine team seemed to agree it was less error prone

Reply:
I don't see why serialization is worse than templating, if not on par?

Reply:
I can change it if you really want, but what does that mean in terms of AKS vs AKS-engine? do we diverge forever? Do I cherry pick that back into AKS-engine? Really want to clarify the process here.

Reply:
@xuto2 why should this be a template snippet instead? I think @alexeldeib makes a good point about reducing potential errors.

Reply:
my only intention is to be consistent with current template approach and don't want to have mixed implementations. My question here is trying to understand why template is not possible, not saying we cannot use serialization.